hs_access_control/models.py

```diff
@@ -129,104 +250,1088 @@ def __str__(self):
                           " over group '{}' (id={})" +
                           " via grantor '{}' (id={})>",
                           str(self.user.username), str(self.user.id),
-                          PrivilegeCodes.PRIVILEGE_NAMES[self.privilege],
+                          PrivilegeCodes.NAMES[self.privilege],
                           str(self.privilege),
                           str(self.group.name), str(self.group.id),
                           str(self.grantor.username), str(self.grantor.id))
 
+    @classmethod
+    def share(cls, **kwargs):
+        """
+        Share a group with a user and update provenance
+
+        ***This completely bypasses access control*** but keeps provenance in sync.
+
+        :param group: target group.
+        :param user: target user.
+        :param privilege: privilege 1-4.
+        :param grantor: user who requested privilege.
+
+        Usage:
+            UserGroupPrivilege.share(group={X}, user={Y}, privilege={Z}, grantor={W}
+        """
+        if __debug__:
+            assert 'group' in kwargs
+            assert isinstance(kwargs['group'], Group)
+            assert 'user' in kwargs
+            assert isinstance(kwargs['user'], User)
+            assert 'grantor' in kwargs
+            assert isinstance(kwargs['grantor'], User)
+            assert 'privilege' in kwargs
+            assert \
+                kwargs['privilege'] >= PrivilegeCodes.OWNER and \
+                kwargs['privilege'] <= PrivilegeCodes.NONE
+            assert len(kwargs) == 4
+        cls.update(**kwargs)
+        UserGroupProvenance.update(**kwargs)
+
+    @classmethod
+    def unshare(cls, **kwargs):
+        """
+        Unshare a group with a user and update provenance
+
+        ***This completely bypasses access control*** but keeps provenance in sync.
+
+        :param group: target group.
+        :param user: target user.
+        :param grantor: user who requested privilege.
+
+        Usage:
+            UserGroupPrivilege.unshare(group={X}, user={Y}, grantor={W})
+
+        Important: this does not guard against removing a single owner.
+
+        **This is a system routine** that should not be called directly by developers!
+        Use UserAccess.unshare_group_with_user instead. That routine avoids single-owner
+        deletion.
+        """
+        if __debug__:
+            assert 'group' in kwargs
+            assert isinstance(kwargs['group'], Group)
+            assert 'user' in kwargs
+            assert isinstance(kwargs['user'], User)
+            assert 'grantor' in kwargs
+            assert isinstance(kwargs['grantor'], User)
+            assert len(kwargs) == 3
+        cls.update(privilege=PrivilegeCodes.NONE, **kwargs)
```

Comment:
I would prefer to be explicit. Not much code, and a matter of personal taste. I think this documents what the code is doing better than the shorter form....